base/base/ReplxxLineReader.cpp

```diff
@@ -293,116 +468,69 @@ void ReplxxLineReader::addToHistory(const String & line)
         rx.print("Unlock of history file failed: %s\n", errnoToString().c_str());
 }
 
-/// See comments in ShellCommand::executeImpl()
-/// (for the vfork via dlsym())
-int ReplxxLineReader::executeEditor(const std::string & path)
+void ReplxxLineReader::openEditor()
 {
-    std::vector<char> argv0(editor.data(), editor.data() + editor.size() + 1);
-    std::vector<char> argv1(path.data(), path.data() + path.size() + 1);
-    char * const argv[] = {argv0.data(), argv1.data(), nullptr};
+    TemporaryFile editor_file("clickhouse_client_editor_XXXXXX.sql");
+    editor_file.write(rx.get_state().text());
+    editor_file.close();
 
-    static void * real_vfork = dlsym(RTLD_DEFAULT, "vfork");
-    if (!real_vfork)
+    char * const argv[] = {editor.data(), editor_file.getPath().data(), nullptr};
+    try
     {
-        rx.print("Cannot find symbol vfork in myself: %s\n", errnoToString().c_str());
-        return -1;
-    }
-
-    pid_t pid = reinterpret_cast<pid_t (*)()>(real_vfork)();
-
-    if (-1 == pid)
-    {
-        rx.print("Cannot vfork: %s\n", errnoToString().c_str());
-        return -1;
+        if (executeCommand(argv) == 0)
+        {
+            const std::string & new_query = readFile(editor_file.getPath());
+            rx.set_state(replxx::Replxx::State(new_query.c_str(), new_query.size()));
+        }
     }
-
-    /// Child
-    if (0 == pid)
+    catch (const std::runtime_error & e)
     {
-        sigset_t mask;
-        sigemptyset(&mask);
-        sigprocmask(0, nullptr, &mask); // NOLINT(concurrency-mt-unsafe) // ok in newly created process
-        sigprocmask(SIG_UNBLOCK, &mask, nullptr); // NOLINT(concurrency-mt-unsafe) // ok in newly created process
-
-        execvp(editor.c_str(), argv);
-        rx.print("Cannot execute %s: %s\n", editor.c_str(), errnoToString().c_str());
-        _exit(-1);
+        rx.print(e.what());
     }
 
-    int status = 0;
-    do
-    {
-        int exited_pid = waitpid(pid, &status, 0);
-        if (exited_pid == -1)
-        {
-            if (errno == EINTR)
-                continue;
-
-            rx.print("Cannot waitpid: %s\n", errnoToString().c_str());
-            return -1;
-        }
-        else
-            break;
-    } while (true);
-    return status;
+    if (bracketed_paste_enabled)
+        enableBracketedPaste();
 }
 
-void ReplxxLineReader::openEditor()
+void ReplxxLineReader::openInteractiveHistorySearch()
 {
-    char filename[] = "clickhouse_replxx_XXXXXX.sql";
-    int fd = ::mkstemps(filename, 4);
-    if (-1 == fd)
+    TemporaryFile history_file("clickhouse_client_history_in_XXXXXX.bin");
```

Comment:
```suggestion
    assert(!fuzzy_finder.empty());
    TemporaryFile history_file("clickhouse_client_history_in_XXXXXX.bin");
```

Reply:
I don't see a reason in this `assert`, since this is not a logical error to put an assert for this, and besides it will not be binded if there is no fzf/sk. Plus nothing wrong will happen except for `No such file or directory` error during `execve`.

But I don't have any objections, so please feel free to commit this suggestion.

Reply:
> besides it will not be binded if there is no fzf/sk.

I see it as a check that the above statement is true, we shouldn't call this function if we didn't find fzf/sk. Right now it's simple, but don't see harm in explicitly having this assumption in assert.